polkadot/node/core/approval-voting/src/lib.rs

```diff
@@ -3163,7 +3174,50 @@ where
 		}
 		if newly_approved {
 			state.record_no_shows(session_index, para_id.into(), &status.no_show_validators);
+			let mut collected_useful_approvals = vec![];
+
+			match status.required_tranches {
+				RequiredTranches::All => {
+					collected_useful_approvals.extend(candidate_entry.approvals().iter_ones().into());
+				},
+				RequiredTranches::Exact {needed, ..} => {
+					let tranches = approval_entry.tranches();
+					for validator_approval_idx in candidate_entry.approvals().iter_ones() {
+						for tranche_entry in tranches {
+							if tranche_entry.tranche() <= needed {
+								let trance_assignments = tranche_entry.assignments();
+								let useful_vote = trance_assignments
+									.iter()
+									.find(|(validator_on_tranche, _)| *validator_on_tranche == ValidatorIndex(validator_approval_idx as _));
```

Comment:
This looks correct, but pretty heavy you have 3 imbricated loops(with the find), given this is on the hot-path I would try to make it more efficient, what about something like:

```
		let mut assigned_mask = approval.assignments_up_to(needed);
		let approvals = candidate.approvals();

		assigned_mask &= approvals;

      collected_useful_approvals =  assigned_mask.iter_ones().collect();
```